### PR TITLE
Read exports from side modules at link time

### PR DESCRIPTION
**Pull request by sbc100**
This means we can reason about the symbols which are or are not defined across 
the entire program which allows for `ERROR_ON_UNDEFINED_SYMBOLS` to be
enabled for any build that includes all its side modules on the command line.

### CODE REVIEW

File: src/compiler.js [hunk 1]
```diff
@@ -67,6 +67,7 @@ if (settingsFile) {
 EXPORTED_FUNCTIONS = set(EXPORTED_FUNCTIONS);
 EXCEPTION_CATCHING_ALLOWED = set(EXCEPTION_CATCHING_ALLOWED);
 IMPLEMENTED_FUNCTIONS = set(IMPLEMENTED_FUNCTIONS);
+SIDE_MODULE_SYMBOLS = set(SIDE_MODULE_SYMBOLS);
```
Comment (dschuff) [inline src/compiler.js:70]: There is a mix of terminology here, e.g. implemented vs exported (vs defined, which is probably the "correct" thing as it applies to linking symbols), as well as functions vs symbols, etc. Are functions and symbols equivalent in this context? Are exported and defined equivalent in this context?

Comment (sbc100) [reply src/compiler.js:70]: I have another PR that renames `IMPLEMENTED_FUNCTIONS` to `WASM_EXPORTS` so you are right we should probably call this `SIDE_MODULE_EXPORTS` which is a more accurate description.

Comment (sbc100) [reply src/compiler.js:70]: If we can land this change first then the names won't appear so different: https://github.com/emscripten-core/emscripten/pull/13917

File: tools/webassembly.py [hunk 2]
```diff
@@ -133,7 +133,8 @@ class ExternType(IntEnum):
 
 Section = namedtuple('Section', ['type', 'size', 'offset'])
 Limits = namedtuple('Limits', ['flags', 'initial', 'maximum'])
-Import = namedtuple('Import', ['type', 'mod', 'field'])
+Import = namedtuple('Import', ['kind', 'mod', 'field'])
```
Comment (dschuff) [inline tools/webassembly.py:136]: seems like "module" would be more readable than "mod" and not really that much less to type?

**Review by dschuff — APPROVED** [review]
Can you also update the documentation? Once this works (is that with this PR?) we should document it as the primary way to create dynamic libraries (btw does this mean we don't need the MAIN_MODULE and SIDE_MODULE flags anymore?)

**Comment by sbc100** [comment]
> (btw does this mean we don't need the MAIN_MODULE and SIDE_MODULE flags anymore?)

Not yet but that would be really really cool.. lets do it!

- MAIN_MODULE -> include a dylib on the command line when linking
= SIDE_MODULE -> `-shared`

**Comment by sbc100** [comment]
Added a ChangeLog entry.

**Comment by dschuff** [comment]
> > (btw does this mean we don't need the MAIN_MODULE and SIDE_MODULE flags anymore?)
> 
> Not yet but that would be really really cool.. lets do it!
> 
> * MAIN_MODULE -> include a dylib on the command line when linking
> * SIDE_MODULE -> `-shared`

I wonder what the least "surprising" way to do MAIN_MODULE is. on other platforms, dynamic executables are the default, but not for wasm. If someone or their build system is including libs on the command line that we are ignoring now, I guess someone could accidentally end up with a dynamic executable? I guess that doesn't seem too likely though, since they would also have to be using SIDE_MODULE.
We should probably also check the behavior of CMake and see if our CMake support needs to be updated.

**Comment by sbc100** [comment]
> on other platforms, dynamic executables are the default, but not for wasm.

Actually on linux (gcc/clang, etc) the behaviour is "make a static executable unless there is one ore more dynamic libraries in the link, in which case make a dynamic executable".      You don't actually need `-static` unless you have shared objects in your libraries path that you want to explicitly ignore and force the use of the static libraries.  If the linker only sees static libraries it will build a static executable.  

So I guess I'm proposing that we match this behaviour.. that major difference being that we don't provide any shared libraries for system libs.. so most users will end up with static by default.

**Comment by sbc100** [comment]
This was surprising to me actually:

```
$ cat test.c
void _start() {}
$ clang -nostdlib test.c
$ file a.out 
a.out: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=94968d2d1c7dfd08c5b9c950ad85d6ae4d42a410, not stripped
```

Here we see the compiler defaulting to static linking (without any `-static`)

**Comment by dschuff** [comment]
OK, you got me there :D 
Still, the dynamic libc in basically every distro means that every "normal" build with no special flags is dynamic.

**Comment by dschuff** [comment]
BTW, I think that your proposal will probably be fine. But we should make sure that CMake is doing something reasonable in the default cases, and maybe it would be good to have some easy way for a user to tell whether an executable is dynamic, to help them (and us) diagnose problems when things go wrong.

**Pull request merged**